src/edu/buffalo/cse/jive/finiteStateMachine/parser/expression/temporal/EExpression.java

```diff
@@ -22,7 +30,84 @@ public EExpression(Expression expression) {
 
 	@Override
 	public Boolean evaluate(Context context) {
-		return evaluate(null, context.getCurrentState(), new HashSet<State>(), context.getStates());
+		
+		if(context.isUseMarker())return evaluate(null, context.getCurrentState(), context.getStates());
+		return evaluate2(null, context.getCurrentState(), new HashSet<State>(), context.getStates());
+		
+	}
+
+	/**
+	 * EF Expression Implementation - Traverses the whole graph and returns if any
+	 * state is true. Marks state as invalid if all child states are invalid.
+	 * Finds the shortest path to the target
+	 * 
+	 * @param prev
+	 * @param curr
+	 * @param visited
+	 * @param states
+	 * @return
+	 */
+	private Boolean evaluate(State prev, State rootState, Map<State, Set<State>> states) {
+		boolean currentResult = false;
+		Queue<State> toBeVisited = new LinkedList<State>();
+		Stack<State> visited = new Stack<State>();
+		toBeVisited.add(rootState);
+		visited.add(rootState);
+		State targetState = new State();
+		while(!toBeVisited.isEmpty()){
+			State curr = toBeVisited.poll();
+			for (State next : states.get(curr)){
+				if(!visited.contains(next)) {
+					boolean isMatch = getExpression().evaluate(new Context(next, null, states));
+					if(isMatch) {
+						targetState = next;
+						currentResult = true;
+						break;	
+					}
+					visited.push(next);
+					toBeVisited.add(next);
+				}
+			}	
+			if(currentResult)break;
+		}
+		
+		if(!currentResult) return currentResult;	
+		List<State> shortestPathList = constuctShortestPath(targetState, visited, states);
+		
+		
+		markStatesInPath(shortestPathList, states);
+		
+		return currentResult;
+	}
+	

```

Comment:
Using BFS traversal we try to find the state which matches with the given E property. If it finds the target state, it uses the _constuctShortestPath_ method to find the path from root and save it in the static class ShortestPathHolderForEExpression for later use.